### DISCUSSION TITLE
Welcome to assembly-evm-opcodes-and-formal-verification-course Discussions!

### DISCUSSION THREAD

**Discussion by PatrickAlphaC** [Cyfrin/assembly-evm-opcodes-and-formal-verification-course#1] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by EngrPips** [comment]
It's about to get interesting!!!!!!!!!!!!!!!!

**Comment by Renzo1** [comment]
Please what could be the possible cause of this error I keep getting whenever I try a halmos test: halmos --function check_testFunctionName

```
[⠒] Compiling...

[⠒] Compiling 21 files with 0.8.22

[⠰] Solc 0.8.22 finished in 3.44s

Compiler run successful!

Skipped stdStorage.json due to parsing failure: KeyError: 'ast'

Skipped stdStorageSafe.json due to parsing failure: KeyError: 'ast'

Skipped Test.json due to parsing failure: KeyError: 'ast'

Skipped StdInvariant.json due to parsing failure: KeyError: 'ast'

Skipped StdStyle.json due to parsing failure: KeyError: 'ast'

Skipped DSTest.json due to parsing failure: KeyError: 'ast'

Skipped Vm.json due to parsing failure: KeyError: 'ast'

Skipped VmSafe.json due to parsing failure: KeyError: 'ast'

Skipped StdAssertions.json due to parsing failure: KeyError: 'ast'

Skipped CommonBase.json due to parsing failure: KeyError: 'ast'

Skipped TestBase.json due to parsing failure: KeyError: 'ast'

Skipped ScriptBase.json due to parsing failure: KeyError: 'ast'

Skipped stdJson.json due to parsing failure: KeyError: 'ast'

Skipped console.json due to parsing failure: KeyError: 'ast'

Skipped Base_Test.json due to parsing failure: KeyError: 'ast'

Skipped stdMath.json due to parsing failure: KeyError: 'ast'

Skipped safeconsole.json due to parsing failure: KeyError: 'ast'

Skipped StdChains.json due to parsing failure: KeyError: 'ast'

Skipped StdUtils.json due to parsing failure: KeyError: 'ast'

Skipped console2.json due to parsing failure: KeyError: 'ast'

Skipped IMulticall3.json due to parsing failure: KeyError: 'ast'

Skipped StdCheatsSafe.json due to parsing failure: KeyError: 'ast'

Skipped StdCheats.json due to parsing failure: KeyError: 'ast'

Skipped MathMastersTest.json due to parsing failure: KeyError: 'ast'

Skipped MathMasters.json due to parsing failure: KeyError: 'ast'

Skipped stdError.json due to parsing failure: KeyError: 'ast'

Error: No tests with the prefix `check_testMulWadUpFuzz`

```

I am following the tutorial and have double-checked everything, but I am unable to successfully run any halmos test.

**Reply by Renzo1** [reply]
Never mind. 
For anyone experiencing this issue, remove your current installation, and reinstall it with pipx.

**Reply by EngrPips** [reply]
Glad you figured it out man. Keep crushing it !

**Comment by Renzo1** [comment]
My halmos test keeps killing itself (timing out) despite setting it to run indefinitely. In the 33rd video of the MathMaster section, where we are expected to catch a bug using Halmos, my test keeps timing out after approximately 5 minutes despite setting the -solver-timeout-assertion flag to 0.

I expected it to run until it found the bug or I killed the run manually, as explained by Patrick, but it kept killing itself, even when I bumped up the timer to a high value like 10000000.

command
```
halmos --function check_testMulWadUpFuzz --solver-timeout-assertion 0

or 


halmos --function check_testMulWadUpFuzz --solver-timeout-assertion 10000000
```

Output
```
[⠔] Compiling...
[⠢] Compiling 21 files with 0.8.22
[⠑] Solc 0.8.22 finished in 6.80s
Compiler run successful!

Running 1 tests for test/MathMasters.t.sol:MathMastersTest
Killed
```

Please what could be the possible reason?

Environment:
- OS: Linux Ubuntu
- Python version: Python 3.12.2
- Halmos and other dependency versions: Halmos 0.1.11

**Reply by PatrickAlphaC** [reply]
Hmm... This might be an issue with the version of python? I'm not sure... could you make this issue on the halmos github?

**Reply by Renzo1** [reply]
For anyone experiencing thesame issue check this link: https://github.com/a16z/halmos/issues/261#issue-2207830885

Thank you, Patrick

**Reply by EngrPips** [reply]
Awesome, thanks for sharing friend

**Reply by Renzo1** [reply]
My pleasure

**Comment by omarhibshi** [comment]
Hello,
I noticed that lessons 18 and 19 have the same video which is the one for lesson 18 (Dispatching). 
The video for (Opcode EQ) is missing.

Thanks

**Reply by EngrPips** [reply]
Thanks for reporting this

**Reply by omarhibshi** [reply]
you welcome!

**Comment by FeeqLabs** [comment]
I am having this error- CRITICAL: Encountered an error running Certora Prover, please contact Certora

whenever i try runnig this command - `certoraRun ./test/invariant-break/formal-verification/certora/conf/FVCatches.conf`

Please what could be the issue and possible solution?

**Reply by EngrPips** [reply]
I can't say what is wrong, but you should check Certora GitHub. Maybe a similar issue was reported already, and what was the solution? While we wait and hope that some intelligent people see this and help

**Comment by JJtheAndroid** [comment]
Question on lesson 18 section 3. Patrick uses DISPACTER(true) to help connect multiple contracts, but the certora docs uses link in the config files. I have been using the latter but I wanted to know which method is better

**Reply by EngrPips** [reply]
I am unsure which is better, but @PatrickAlphaC will let you know when he is available.

**Reply by PatrickAlphaC** [reply]
Sorry on my delay. I think you're [referring to this](https://docs.certora.com/projects/tutorials/en/latest/lesson4_invariants/ghosts/exercises.html#english-auction). It just depends on what you're going for.

1. If you're unsure about the files/projects/have a lot of projects you're going to use, `DISPATCHER` might be easier
2. If you have storage slots that you are sure are going to be specific contracts, you can use `link` in the config.

Just know, both options technically make your formal verification unsound. But yes, both are fine, it just depends on what you're going for.

**Comment by martinezjorge** [comment]
When I try running

```
forge test --match-path *Huff* --debug testReadValue -vvvv 
```

I get the following output

```
zsh: no matches found: *Huff*
```

I'm not sure if its zsh acting weird or my terminal configuration.

**Reply by EngrPips** [reply]
Hello @martinezjorge, I was getting the same output until I manually declare the path.

```js
forge test --match-path /test/... --debug testReadValue -vvvv
```

**Reply by martinezjorge** [reply]
Thanks for the response Pips,

I was able to get the debugger to run and isolate the test using the following command

```
forge test --match-path ./test/v1/HorseStoreHuff.t.sol --debug testReadValue
```

What made the biggest difference for me was putting the period in front the file path

**Reply by EngrPips** [reply]
Awesome, Glad you resolved this issue. Keep crushing it. Friend.

**Comment by Falilah** [comment]
Hi all, I upgraded my certora, then I started getting this error when I was practicing the `hook and ghost` concept from the gasBadNFT course.
In this error, certora is saying it can't see the private state variable(mapping) from the current contract whereas the variable is declared in the contract.
```ERROR: Failed to compile spec files:
CRITICAL: [main] ERROR ALWAYS - Found errors in certora/spec/GasBad.spec:
CRITICAL: [main] ERROR ALWAYS - Error in spec file (GasBad.spec:13:1): named pattern root 's_listings' is not defined: did you spell something wrong?
CRITICAL: Encountered an error running Certora Prover:```
have tried researching on why such can occur but I could not get anything to solve it

**Reply by EngrPips** [reply]
Hello @Falilah, I have no knowledge of what could cause this but maybe @PatrickAlphaC  knows what could be wrong. They will see to this as soon as they are available.

**Reply by PatrickAlphaC** [reply]
What do you mean you "upgraded your certora"? 

Could you share your spec here?

**Reply by Falilah** [reply]
sure
I updated the version to the newest version.
here is my spec file
```spec
ghost mathint listingUpdatesCounts{
    init_state axiom listingUpdatesCounts == 0;
}
ghost mathint log4Count{
    init_state axiom log4Count == 0;
}


hook Sstore s_listings[KEY address nftAddress][KEY uint256 tokenId].price uint256 price {
    listingUpdatesCounts = listingUpdatesCounts + 1;
}
hook LOG4(uint offset, uint length, bytes32 t1, bytes32 t2, bytes32 t3, bytes32 t4){
    log4Count = log4Count + 1;
}

/* //////////////////////////////////////////////////////////////
        RULES
  ///////////////////////////////////////////////////////////// */

  invariant anytime_mapping_updated_emit_event()
    listingUpdatesCounts <= log4Count;
```

**Reply by PatrickAlphaC** [reply]
Sorry on my delay. It would seem to me that `s_listings` isn't in your solidity. Could you try renaming the variable perhaps?

**Comment by cyfe45** [comment]
If anyone is looking for the link to Alejandro Santander Deconstruction a Solidity Contract series [here](https://medium.com/@_ajsantander) is the link.

**Reply by EngrPips** [reply]
Thanks for this heads-up buddy.
